docs/features/shared-volumes-with-NFS.md

```diff
@@ -0,0 +1,129 @@
+
+## Requirements
+* Developer should able to use externally managed NFS share as “shared docker volume” using “vsphere” volume driver.
+* This allows them to use NFS directories as docker volumes using Docker APIs, Compose etc 
+* Extend the vSphere Docker Volume plugin to support NFS shared volumes and mutliple readers/writers access
+* Volumes are accessible on multiple docker hosts running on  multiple ESX hosts in a Docker swarm
+* Allow admin to specify what remote dirs are managed by the plugin
+* Support all current docker volume plugin API for shared volumes
+
+## User experience
+### Set up remote directories for docker volumes
+The design here assumes that the admin is responsible to identify remote directories to export and has configured remote servers to export those. Specifically, the admin needs to have,
+1. Exported remote directories to be used to host docker volumes.
+1. Configured sufficient permissions for the admins on the Docker host to be able to mount and create folders and files on those directories.
+1. The docker host has the necessary packages of the desired version of NFS installed.
+1. The admin can use the same remote directory configuration on multiple docker hosts on separate physical hosts and be able to view and use the same volumes on their containers (the applications must be able to synchronize access to shared volumes).
+
+### Update plugin configuration
+The plugin needs to be provided the details of the exported diretories and the servers from which to mount them. The details are entered into the plugin configuration file at /etc/docker-volume-vsphere.conf in the format as detailed later.
+
+### Restart plugin
+The plugin needs to be restarted to load the updated configuration.
+
+### Using shared volumes
+Docker volume commands must be used as shown below to use volumes on remote directories. Similar to how local VMDK based volumes are qualified with a datastore name, shared volumes must also be qualified with the label (see remote directory configuration below) that identifies the remote dir. to use to locate and refer a volume.
+1. docker volume create - 
+   1. _docker volume create --driver=vsphere --name=remote-vol -o type=<fstype - nfs/nfs4> - uses the default entry in the list of remote dirs to locate the volume
+   1. _docker volume create --driver=vsphere --name=remote-vol@remote-dir-label_ - creates the volume in the location identified by the label in the remote dirs list
+1. docker volume list - No change
+1. docker volume inspect - the volume must be qualified by the remote dir label as below,
+   1. _docker volume inspect vol-name@remote-dir-label_
+1. docker run - the volume must be qualified by the remote dir label as below,
+   1. _docker run -it --volume-driver=vsphere -v remote-vol@remote-dir-label:/vol1 --name ub ubuntu_
+1. docker volume rm - the volume must be qualified by the remote dir label as below,
+   1. _docker volume rm vol-name@remote-dir-label_
```

Comment:
Its exactly as how today if a volume is created with a datastore name (or label) then it needs to be specified as such in all the docker commands. As below.

_docker volume create --driver=vsphere --name=TestVol60@bigone -o size=400mb_

_docker volume inspect TestVol60@bigone_
_[
    {
        "Driver": "vsphere:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vmdk/TestVol60@bigone",
        "Name": "TestVol60@bigone",
        "Options": {
            "size": "400mb"
        },_
_....._

_docker run -it --volume-driver=vsphere -v Test_Vol60@bigone:/vol1 --name ub1 ubuntu
root@9c842f9aada7:/#
root@9c842f9aada7:/#_

With NFS the proposal is to have a default and gets used on create. But after that on all commands the user needs to specify the volume name as volname@label. This is to ensure that the driver can identify which remote share hosts the volume. The same volume name can be on multiple remote shares, Hence user specifies the label.

Even with VMDK backed volumes I can create ABC@DS1 and later if I specify ABC@DS1 the plugin will pick the correct volume. But if I specify just ABC on the docker command then either the plugin will check the default DS and report _volume not found_ or worse pick the wrong volume.

Regarding accessing and removing a volume on different hosts, prefer the plugin doesn't keep meta-data and the need to maintain it. The best justification is this - the volume plugin isn't cluster aware as yet and will have to rely on Docker doing the _right thing_ when using volumes in a swarm cluster. I'll update the doc.